### PR TITLE
Minimize autojoins in dimensional 1-2 cases

### PR DESCRIPTION
**Pull request by djspiewak**
Example:

```sql
select a[*] + b from foo
```

This is pretty narrowly tailored, but it does address part of #3222

### CODE REVIEW

**Comment by djspiewak** [comment]
This is going to fail on appveyor because of cache invalidation issues (zinc has a bug w.r.t. the exact change that I made).

**Comment by djspiewak** [comment]
Even aside from the build failures, don't merge this yet.  Reduction coalescence will be broken in some weird cases by this change.

**Comment by djspiewak** [comment]
Ok correction: it doesn't break reduction coalescence, and thus can be merged.  It's just that recursive coalescence is quite naive right now, which results in missing optimization chances.

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 1]
```diff
@@ -159,13 +160,25 @@ final class MinimizeAutoJoins[T[_[_]]: BirecursiveT: EqualT] private () extends
           coalesceToMap[G](qgraph, candidates, fm)
 
         case None =>
-          val reducerAttempt = candidates collect {
+          lazy val reducerAttempt = candidates collect {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: `reducerCheck`, forced on L181, will in turn force this, so does the `lazy` actually provide any benefit?

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 2]
```diff
@@ -159,13 +160,25 @@ final class MinimizeAutoJoins[T[_[_]]: BirecursiveT: EqualT] private () extends
...
             case g @ QSReduce(source, buckets, reducers, repair) =>
               (source, buckets, reducers, repair)
           }
 
           // candidates.forall(_ ~= QSReduce)
-          if (reducerAttempt.lengthCompare(candidates.length) === 0) {
+          lazy val reducerCheck = reducerAttempt.lengthCompare(candidates.length) === 0
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Always forced on L181.

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 3]
```diff
@@ -253,28 +266,35 @@ final class MinimizeAutoJoins[T[_[_]]: BirecursiveT: EqualT] private () extends
...
-                  }
+                  _ <- updateForCoalesce[G](candidates, red.root)
                 } yield back
               } else {
                 failure[G](qgraph)
               }
             } yield back
           } else {
-            failure[G](qgraph)
+            leftShift12Extract match {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Looks like we need to start thinking about how to factor these transformations out, this is ok for now but not going to scale as we need add more.

**Comment by wemrysi** [comment]
@djspiewak did you intentionally tag this with `version: feature`?

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala (hunk 1 above)
Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Probably not.  This is gone on my follow-up branch.

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala (hunk 3 above)
Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Way ahead of you.  :-)  And I totally agree.

**Comment by djspiewak** [comment]
@wemrysi I did.  Not sure exactly if it makes more sense as a `revision`.  It is optimizing a whole class of queries in a user-visible way.  Does that count as a feature?

**Comment by wemrysi** [comment]
@djspiewak I don't recall if we've labeled optimizations as features in the past, but I'm not opposed to the idea.

**Pull request merged**